src/descriptors.js

```diff
@@ -0,0 +1,150 @@
+  }
+  mightHaveBeenDeleted(): boolean {
+    invariant(false, "should have been overridden by subclass");
+  }
+}
+
+export type DescriptorInitializer = {|
+  writable?: boolean,
+  enumerable?: boolean,
+  configurable?: boolean,
+
+  value?: Value,
+
+  get?: UndefinedValue | CallableObjectValue | AbstractValue,
+  set?: UndefinedValue | CallableObjectValue | AbstractValue,
+|};
+
+// Normal descriptors are returned just like spec descriptors
+export class PropertyDescriptor extends Descriptor {
+  writable: void | boolean;
+  enumerable: void | boolean;
+  configurable: void | boolean;
+
+  // If value instanceof EmptyValue, then this descriptor indicates that the
+  // corresponding property has been deleted.
+  value: void | Value;
+
+  get: void | UndefinedValue | CallableObjectValue | AbstractValue;
+  set: void | UndefinedValue | CallableObjectValue | AbstractValue;
+
+  constructor(desc: DescriptorInitializer | PropertyDescriptor) {
+    super();
+    if (desc.hasOwnProperty("writable")) this.writable = desc.writable;
+    if (desc.hasOwnProperty("enumerable")) this.enumerable = desc.enumerable;
+    if (desc.hasOwnProperty("configurable")) this.configurable = desc.configurable;
+    if (desc.hasOwnProperty("value")) this.value = desc.value;
+    if (desc.hasOwnProperty("get")) this.get = desc.get;
+    if (desc.hasOwnProperty("set")) this.set = desc.set;
+  }
+
+  throwIfNotConcrete(realm: Realm): PropertyDescriptor {
+    return this;
+  }
+  mightHaveBeenDeleted(): boolean {
+    if (this.value === undefined) return false;
+    return this.value.mightHaveBeenDeleted();
+  }
+}
+
+// Only internal properties (those starting with $ / where internalSlot of owning property binding is true) will ever have array values.
+export class InternalSlotDescriptor extends Descriptor {
+  value: void | Value | Array<any>;
+
+  constructor(value?: void | Value | Array<any>) {
+    super();
+    this.value = Array.isArray(value) ? value.slice(0) : value;
+  }
+
+  mightHaveBeenDeleted(): boolean {
+    return false;
+  }
+}
+
+// Only used if the result of a join of two descriptors is not a data descriptor with identical attribute values.
+// When present, any update to the property must produce effects that are the join of updating both desriptors,
+// using joinCondition as the condition of the join.
+export class AbstractJoinedDescriptor extends Descriptor {
+  joinCondition: AbstractValue;
+  descriptor1: void | Descriptor;
+  descriptor2: void | Descriptor;
+
+  constructor(joinCondition: AbstractValue, descriptor1?: Descriptor, descriptor2?: Descriptor) {
+    super();
+    this.joinCondition = joinCondition;
+    this.descriptor1 = descriptor1;
+    this.descriptor2 = descriptor2;
+  }
+  mightHaveBeenDeleted(): boolean {
+    if (this.descriptor1 && this.descriptor1.mightHaveBeenDeleted()) {
+      return true;
+    }
+    if (this.descriptor2 && this.descriptor2.mightHaveBeenDeleted()) {
+      return true;
+    }
+    return false;
```

Comment:
In a way, the absence of either descriptor indicates that the thing doesn't exists that thus `mightHaveBeenDeleted`.